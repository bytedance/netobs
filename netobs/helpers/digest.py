# Copyright 2022-2024 Bytedance Ltd. and/or its affiliates
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from jax import numpy as jnp


def iqr_clip_real(x: jnp.ndarray, iqr_range: float = 3.0) -> jnp.ndarray:
    """Clip outliers from real-valed array with interquartile range (IQR).

    Args:
        x: array to be clipped (must be real)
        iqr_range: how many times the data outside the IQR will be clipped

    Returns:
        Clipped array
    """
    q1 = jnp.quantile(x, 0.25, axis=0, keepdims=True)
    q3 = jnp.quantile(x, 0.75, axis=0, keepdims=True)
    iqr = q3 - q1
    cutoff = iqr_range * iqr
    return jnp.clip(x, q1 - cutoff, q3 + cutoff)


def iqr_clip(x: jnp.ndarray, iqr_range: float = 3.0) -> jnp.ndarray:
    """Clip outliers with interquartile range (IQR).

    Args:
        x: array to be clipped
        iqr_range: how many times the data outside the IQR will be clipped

    Returns:
        Clipped array
    """
    if jnp.iscomplexobj(x):
        return iqr_clip_real(x.real, iqr_range) + 1j * iqr_clip_real(x.imag, iqr_range)
    return iqr_clip_real(x, iqr_range)


def robust_mean(x: jnp.ndarray, iqr_range: float = 3.0) -> jnp.ndarray:
    """Take the mean after clipping.

    Args:
        x: input array
        iqr_range: how many times the data outside the IQR will be clipped

    Returns:
        Array averaged in the first dim.
    """
    clipped = iqr_clip(x, iqr_range)
    return jnp.nanmean(clipped, axis=0)


def robust_mean_std(
    x: jnp.ndarray, iqr_range: float = 3.0
) -> tuple[jnp.ndarray, jnp.ndarray]:
    """Take the mean and estimate stderr after clipping.

    Args:
        x: input array
        iqr_range: how many times the data outside the IQR will be clipped

    Returns:
        A tuple of
        - Array averaged in the first dim.
        - Std of the mean.
    """
    clipped = iqr_clip(x, iqr_range)
    if jnp.iscomplexobj(clipped):
        std = jnp.nanstd(clipped.real, axis=0) + 1j * jnp.nanstd(clipped.imag, axis=0)
    else:
        std = jnp.nanstd(clipped, axis=0)
    return (jnp.nanmean(clipped, axis=0), std / jnp.sqrt(clipped.shape[0]))


def weighted_sum(a: jnp.ndarray, weights: jnp.ndarray) -> jnp.ndarray:
    """Taking average, but align array and weights.

    Ftuple[jnp.ndarray, jnp.ndarray]:tuple[jnp.ndarray, jnp.ndarray]:or example, (ijkl) * (ij) -> (kl).

    Args:
        a: JAX array.
        weights: the weight for averaging.

    Returns:
        sum of weighted array
    """
    axis = tuple(range(len(weights.shape)))
    weights = weights.reshape(weights.shape + (1,) * (a.ndim - weights.ndim))
    return jnp.mean(a * weights, axis=axis)
