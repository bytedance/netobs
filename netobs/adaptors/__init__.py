# Copyright 2022-2024 Bytedance Ltd. and/or its affiliates
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""This file holds common types and interfaces for networks.

All networks need to implement or integrate with NetworkAdaptor class to work with this
codebase. For example,

```python
@register_pytree_node_class
class AwesomeNetAdaptor(NetworkAdaptor):
    def call_network(...):
        ...

    # And other methods ...
```

Usually, to restore and evaluate the network, one may need more arguments than those
passed through arguments of the methods. Then `__init__` is the perfect place to
initialize your adaptor with some configurations and many more. There is no restriction
to the `__init__` method, since all the functions only take an instance of adaptor.

And an instance of `AwesomeNetAdaptor` should then be passed to `evaluate_observable`.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Any, Callable, Generic, Protocol, TypeVar

import jax
from jax import numpy as jnp
from typing_extensions import Self

from netobs.helpers.grad import grad_with_system
from netobs.systems import System

X = TypeVar("X", bound=System)
D = TypeVar("D")


class WalkingStep(Protocol[D]):
    """MCMC walking step function."""

    def __call__(
        self,
        key: jnp.ndarray,
        params: jnp.ndarray,
        electrons: jnp.ndarray,
        aux_data: D,
    ) -> tuple[jnp.ndarray, D]:
        pass


class NetworkAdaptor(ABC, Generic[X]):
    """Common interface for different networks.

    Args:
        config: Anything telling the adaptor how to load network config.
            But a string is preferred because it makes life easier for command line.
        args: Command line args to modify config.
    """

    def __init__(self, config: Any, args: list[str]) -> None:
        super().__init__()

    def tree_flatten(self) -> tuple[tuple, Self]:
        """Flatten the class as PyTree.

        Since the adaptor is usually static and doesn't store any data, we just return
        empty `children` list and `self` as `aux_data`.

        Useful when JIT-ting and `pmap`-ping some methods of the adaptor class.

        Returns:
            A tuple of children (which itself is an empty tuple) and the adaptor itself.
        """
        return ((), self)

    @classmethod
    def tree_unflatten(cls, aux_data: Self, children: tuple) -> Self:
        """Unflatten the class as PyTree.

        Args:
            aux_data: the adaptor itself
            children: ignored

        Returns:
            The adaptor it self.
        """
        del children
        return aux_data

    @abstractmethod
    def restore(self, restore_option: Any) -> tuple[Any, jnp.ndarray, X, Any]:
        """Restore checkpoint with given options.

        Args:
            restore_option: anything needed for restoring the network from checkpoint.
                It's just the `options.network_restore` passed to `evaluate_observable`.

        Returns:
            Restored data:
            - params
            - system
            - aux_data
        """

    @abstractmethod
    def call_signed_network(
        self, params: jnp.ndarray, electrons: jnp.ndarray, system: X
    ) -> tuple[jnp.ndarray, jnp.ndarray]:
        """Get the sign and log magnitude of the wavefunction.

        If the network is complex-valued, return log phase and log psi.

        Args:
            params: network parameters.
            electrons: electron positions, shape (nelectrons*ndim), where ndim is the
                dimensionality of the system.
            system: system containing atomic info.

        Returns:
            The sign (or phase) and log magnitude of the wavefunction.
        """

    def call_network(
        self, params: jnp.ndarray, electrons: jnp.ndarray, system: X
    ) -> jnp.ndarray:
        """Get the log magnitude of the wavefunction.

        Neither sign nor phase factor should be included.

        Args:
            params: network parameters.
            electrons: electron positions, shape (nelectrons*ndim), where ndim is the
                dimensionality of the system.
            system: system containing atomic info.

        Returns:
            The log magnitude of the wavefunction.
        """
        return self.call_signed_network(params, electrons, system)[1]

    @abstractmethod
    def make_walking_step(
        self,
        batch_log_psi: Callable[[jnp.ndarray, jnp.ndarray, X], jnp.ndarray],
        steps: int,
        system: X,
    ) -> WalkingStep:
        """Make MCMC walking step.

        Args:
            batch_log_psi: Batched function returning log psi.
                (params, (nbatch, nelec*ndim), system) -> (nbatch,)
                We receive it as input because the importance sampling can be based on
                another psi_G instead of base psi_T.
            steps: Number of MCMC steps to run.
            system: system containing atomic info.

        Returns:
            Function with the same signature as WalkingStep
        """

    def make_burnin_step(
        self,
        batch_log_psi: Callable[[jnp.ndarray, jnp.ndarray, X], jnp.ndarray],
        steps: int,
        system: X,
    ) -> WalkingStep:
        """Make MCMC burnin step.

        By default, this is the same as make_walking_step.

        Args:
            batch_log_psi: Batched function returning log psi.
            steps: Number of MCMC steps to run.
            system: system containing atomic info.

        Returns:
            Function with the same signature as WalkingStep
        """
        return self.make_walking_step(batch_log_psi, steps, system)

    def call_local_energy(
        self, params: jnp.ndarray, key: jnp.ndarray, electrons: jnp.ndarray, system: X
    ) -> jnp.ndarray:
        """Get local energy.

        Args:
            params: Network params.
            key: JAX random keys.
            electrons: Electon positions.
            system: System containing atomic info.

        Returns:
            Local energy at this point.
        """
        key0, key1 = jax.random.split(key)
        ke = self.call_local_kinetic_energy(params, key0, electrons, system)
        pe = self.call_local_potential_energy(params, key1, electrons, system)
        return ke + pe

    @abstractmethod
    def call_local_kinetic_energy(
        self, params: jnp.ndarray, key: jnp.ndarray, electrons: jnp.ndarray, system: X
    ) -> jnp.ndarray:
        """Get local kinetic energy.

        Args:
            params: Network params.
            key: JAX random keys.
            electrons: Electon positions.
            system: System containing atomic info.

        Returns:
            Local kinetic energy at this point.
        """

    @abstractmethod
    def call_local_potential_energy(
        self, params: jnp.ndarray, key: jnp.ndarray, electrons: jnp.ndarray, system: X
    ) -> jnp.ndarray:
        """Get local potential energy.

        Args:
            params: Network params.
            key: JAX random keys.
            electrons: Electon positions.
            system: System containing atomic info.

        Returns:
            Local potential energy at this point.
        """

    def make_network_grad(self, arg: str, jaxfun: Callable = jax.grad):
        """Create gradient function of network.

        Useful when the gradient handling is different in your network.
        Note that if your network is complex-valued, you don't need to implement
        this function since `self.call_network` must be real.

        Args:
            arg: the name of arguments to calculate gradient, e.g. "electrons"/"atoms".
            jaxfun: the type of gradient to take, e.g. `jax.grad`.

        Returns:
            The gradient function.
        """
        return grad_with_system(self.call_network, arg, jaxfun=jaxfun)  # type: ignore

    def make_signed_network_grad(self, arg: str, jaxfun: Callable = jax.grad):
        """Create gradient function of network, taking sign into consideration.

        Useful when the gradient handling is different in your network,
        e.g. the network is complex-valued.

        Args:
            arg: the name of arguments to calculate gradient, e.g. "electrons"/"atoms".
            jaxfun: the type of gradient to take, e.g. `jax.grad`.

        Returns:
            The gradient function.
        """
        return grad_with_system(self.call_network, arg, jaxfun=jaxfun)  # type: ignore

    def make_local_energy_grad(self, arg: str, jaxfun: Callable = jax.grad):
        """Create gradient function of local energy.

        Useful when the gradient handling is different in your network,
        e.g. the network is complex-valued.

        Args:
            arg: the name of arguments to calculate gradient, e.g. "electrons"/"atoms".
            jaxfun: the type of gradient to take, e.g. `jax.grad`.

        Returns:
            The gradient function.
        """
        return grad_with_system(self.call_local_energy, arg, jaxfun=jaxfun)  # type: ignore


# Utility protocols
class LogPsiNet(Protocol):
    def __call__(
        self, params: jnp.ndarray, electrons: jnp.ndarray, system: System
    ) -> jnp.ndarray:
        pass
