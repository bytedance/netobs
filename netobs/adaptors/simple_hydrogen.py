# Copyright 2022-2024 Bytedance Ltd. and/or its affiliates
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from functools import partial
from typing import Any, TypedDict

import jax
from jax import numpy as jnp
from jax.tree_util import register_pytree_node_class

from netobs.adaptors import NetworkAdaptor, WalkingStep
from netobs.systems.molecule import MolecularSystem


class HydrogenAuxData(TypedDict):
    mcmc_width: jnp.ndarray


@register_pytree_node_class
class SimpleHydrogen(NetworkAdaptor):
    """A simple adaptor for hydrogen atom wavefunction. Used for testing."""

    quality: float
    "How good the wavefunction is "

    def __init__(self, config: Any, args: list[str]) -> None:
        super().__init__(config, args)
        self.quality = float(config or 1)

    def restore(
        self, ckpt_file: str | None = None
    ) -> tuple[Any, jnp.ndarray, MolecularSystem, Any]:
        batch_size = int(ckpt_file or 2048)
        key = jax.random.PRNGKey(42)
        kr, kt, kp = jax.random.split(key, 3)
        shape = (1, batch_size, 1)
        r = jax.random.exponential(kr, shape)
        theta = jax.random.uniform(kt, shape, maxval=jnp.pi)
        phi = jax.random.uniform(kp, shape, maxval=2 * jnp.pi)
        data = jnp.concatenate(
            [
                r * jnp.sin(theta) * jnp.cos(phi),
                r * jnp.sin(theta) * jnp.sin(phi),
                r * jnp.cos(theta),
            ],
            axis=-1,
        )

        self.atoms = jnp.array([[0.0, 0.0, 0.0]])
        system = MolecularSystem(
            atoms=self.atoms, charges=jnp.array([1]), spins=(1, 0), ndim=3
        )
        aux_data = HydrogenAuxData(mcmc_width=jnp.array([0.6]))
        return (jnp.array([0]), data, system, aux_data)

    def call_signed_network(
        self, params: jnp.ndarray, electrons: jnp.ndarray, system: MolecularSystem
    ):
        del params
        # _, r_ae = calculate_r_ae(electrons, system)
        # r_ae = r_ae[0, 0, 0]  # 1st electron, 1st atom
        r_ae = jnp.linalg.norm(electrons - system["atoms"])
        return 1, -r_ae * self.quality - r_ae**2 * (1 - self.quality)

    def make_walking_step(
        self,
        steps: int,
        system: MolecularSystem,
    ) -> WalkingStep[HydrogenAuxData]:
        def walk(
            key: jnp.ndarray,
            params: jnp.ndarray,
            electrons: jnp.ndarray,
            aux_data: HydrogenAuxData,
        ) -> tuple[jnp.ndarray, HydrogenAuxData]:
            stddev = aux_data["mcmc_width"]

            def mh_update(_, data):
                x1, key, lp_1 = data
                key, subkey = jax.random.split(key)
                x2 = x1 + stddev * jax.random.normal(subkey, shape=x1.shape)  # proposal
                lp_2 = 2.0 * self.batch_network(params, x2, system)
                ratio = lp_2 - lp_1

                key, subkey = jax.random.split(key)
                cond = ratio > jnp.log(jax.random.uniform(key, shape=ratio.shape))
                x_new = jnp.where(cond[..., None], x2, x1)
                lp_new = jnp.where(cond, lp_2, lp_1)

                return x_new, key, lp_new

            logprob = 2.0 * self.batch_network(params, electrons, system)
            new_data, *_ = jax.lax.fori_loop(
                0, steps, mh_update, (electrons, key, logprob)
            )
            return new_data, aux_data

        return jax.pmap(walk, in_axes=0)

    def call_local_kinetic_energy(
        self,
        params: jnp.ndarray,
        key: jnp.ndarray,
        electrons: jnp.ndarray,
        system: MolecularSystem,
    ) -> jnp.ndarray:
        del key
        n = electrons.shape[0]
        eye = jnp.eye(n)
        grad_f = self.make_network_grad("electrons")

        def grad_f_closure(x):
            return grad_f(params, x, system)

        primal, dgrad_f = jax.linearize(grad_f_closure, electrons)

        return -0.5 * jax.lax.fori_loop(
            0, n, lambda i, val: val + dgrad_f(eye[i])[i], 0.0
        ) - 0.5 * jnp.sum(primal**2)

    def call_local_potential_energy(
        self,
        params: jnp.ndarray,
        key: jnp.ndarray,
        electrons: jnp.ndarray,
        system: MolecularSystem,
    ) -> jnp.ndarray:
        del params, key
        r_ae = jnp.linalg.norm(electrons - system["atoms"])
        # _, r_ae = calculate_r_ae(electrons, system)
        # r_ae = r_ae[0, 0, 0]  # 1st electron, 1st atom
        return -1 / r_ae

    @partial(jax.vmap, in_axes=(None, None, 0, None))
    def batch_network(self, params, electrons, system):
        return self.call_network(params, electrons, system)


DEFAULT = SimpleHydrogen
