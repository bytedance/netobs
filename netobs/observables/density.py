# Copyright 2022-2024 Bytedance Ltd. and/or its affiliates
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Any

from jax import numpy as jnp

from netobs.observables import Estimator, Observable


class Density(Observable):
    def shapeof(self, system) -> tuple[int, ...]:
        return ()


class DensityEstimator(Estimator):
    observable_type = Density

    def __init__(self, adaptor, system, estimator_options, observable_options):
        super().__init__(adaptor, system, estimator_options, observable_options)

        bins = self.options.get("bins", 50)
        if isinstance(bins, int):
            bins = [bins] * system["ndim"]
        self.hist_bins = jnp.array(bins)

        hist_range = self.options.get("range", None)
        if hist_range is None:
            self.hist_range = None
        else:
            self.hist_range = jnp.array(hist_range)

    def empty_val_state(
        self, steps: int
    ) -> tuple[dict[str, jnp.ndarray], dict[str, Any]]:
        del steps
        return {}, {
            "map": jnp.zeros(self.hist_bins),
            "range": self.hist_range,
        }

    def evaluate(
        self, i, params, key, data, system, state, aux_data
    ) -> tuple[dict[str, jnp.ndarray], dict[str, Any]]:
        del i, params, aux_data, key
        x = jnp.reshape(data, (-1, system["ndim"]))
        if state["range"] is None:
            max_coord = jnp.amax(x, axis=0)
            min_coord = jnp.amin(x, axis=0)
            state["range"] = jnp.c_[min_coord, max_coord]
        state["map"] += jnp.histogramdd(x, self.hist_bins, state["range"])[0]
        return {}, state

    def digest(self, all_values, state) -> dict[str, jnp.ndarray]:
        del all_values, state
        return {}


DEFAULT = DensityEstimator  # Useful in CLI
