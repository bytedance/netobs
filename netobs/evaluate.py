# Copyright 2022-2024 Bytedance Ltd. and/or its affiliates
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import dataclasses
import time
from typing import Any, Optional

import jax
from jax import numpy as jnp

from netobs.adaptors import NetworkAdaptor
from netobs.checkpoint import CheckpointManager
from netobs.helpers.digest import robust_mean_std, weighted_sum
from netobs.logging import logger
from netobs.observables import Estimator
from netobs.options import NetObsOptions

# Utilities for paralleled random keys
broadcast_all_local_devices = jax.pmap(lambda x: x)
p_split = jax.pmap(lambda key: tuple(jax.random.split(key)))


def make_different_rng_key_on_all_devices(
    rng: jnp.ndarray,
) -> jnp.ndarray:
    rng = jax.random.fold_in(rng, jax.process_index())
    rng = jax.random.split(rng, jax.device_count())
    return broadcast_all_local_devices(rng)


def evaluate_observable(
    network_adaptor: NetworkAdaptor,
    estimator_class: type[Estimator],
    checkpoint_mgr: Optional[CheckpointManager] = None,
    options: NetObsOptions | None = None,
) -> tuple[dict[str, Any], dict[str, Any], dict[str, Any]]:
    """Run inference to evaluate observable for a system.

    Args:
        network_adaptor: Network adaptor.
        estimator_class: An estimator class. See estimators.py for detail.
        checkpoint_mgr: The checkpoint manager.
        options: The NetObsOptions. See the docstring of NetObsOptions for detail.

    Returns:
        * entries returned by `estimator.digest`
        * values and states returned by `estimator.evaluate`
    """
    if checkpoint_mgr is None:
        checkpoint_mgr = CheckpointManager()
    if options is None:
        options = NetObsOptions()

    params, data, system, aux_data = network_adaptor.restore(options.network_restore)

    key = jax.random.PRNGKey(options.random_seed)

    estimator = estimator_class(
        network_adaptor, system, options.estimator, options.observable
    )

    sharded_key = make_different_rng_key_on_all_devices(key)

    empty_val, empty_state = estimator.empty_val_state(options.steps)

    if options.reweight_ratio > 0.0:
        logger.info("Reweighting is enabled.")
        empty_val["reweighting_weights"] = jnp.zeros((options.steps,))

    init_step, new_data, all_values, state, aux_data = checkpoint_mgr.restore(
        empty_val, empty_state, aux_data
    )
    if new_data is not None:
        data = new_data

    metadata = {
        "version": 9,
        "estimator": estimator_class.__name__,
        **dataclasses.asdict(options),
    }

    if init_step >= options.steps:  # re-exporting data
        digest = estimator.digest(all_values, state)
        log_digest(options.steps - 1, digest)
        if init_step != options.steps:  # save ckpt with less steps
            checkpoint_mgr.save(
                options.steps - 1, data, digest, all_values, state, aux_data, metadata
            )
        return digest, all_values, state

    batch_log_psi = jax.vmap(network_adaptor.call_network, (None, 0, None))
    pmap_log_psi = jax.pmap(batch_log_psi, in_axes=(0, 0, None))

    if options.reweight_ratio > 0.0:
        log_psi = pmap_log_psi(params, data, system)
        log_eps = options.reweight_exp * jnp.quantile(log_psi, options.reweight_ratio)
        logger.info("Setting log epsilon to %s", log_eps)

        def reweighting_log_psi(params, electrons, system):
            log_psi = network_adaptor.call_network(params, electrons, system)
            return jnp.maximum(log_psi, log_eps + (1 - options.reweight_exp) * log_psi)

        batch_guide_log_psi = jax.vmap(reweighting_log_psi, (None, 0, None))
    else:
        batch_guide_log_psi = batch_log_psi

    logger.info("Start burning in %s steps", options.mcmc_burn_in)
    call_burnin_step = network_adaptor.make_burnin_step(
        batch_guide_log_psi, options.mcmc_burn_in * options.mcmc_steps, system
    )
    sharded_key, subkeys = p_split(sharded_key)
    call_burnin_step(subkeys, params, data, aux_data)

    call_walking_step = network_adaptor.make_walking_step(
        batch_guide_log_psi, options.mcmc_steps, system
    )

    time_start = None  # initialize later to be more accurate
    last_log = time.time()
    last_save = time.time()
    logger.info("Starting %s evaluation steps", options.steps - init_step)
    for i in range(init_step, options.steps):
        sharded_key, subkeys = p_split(sharded_key)
        obs_values, state = estimator.evaluate(
            i, params, subkeys, data, system, state, aux_data
        )

        if options.reweight_ratio > 0.0:
            log_psi = pmap_log_psi(params, data, system)
            weights = jnp.minimum(
                1, jnp.exp(options.reweight_exp * 2 * log_psi - 2 * log_eps)
            )
            mean_obs_values = {
                k: weighted_sum(v, weights) for k, v in obs_values.items()
            }
            mean_obs_values["reweighting_weights"] = jnp.mean(weights)
        else:
            mean_obs_values = {k: jnp.mean(v, (0, 1)) for k, v in obs_values.items()}

        all_values = {k: v.at[i].set(mean_obs_values[k]) for k, v in all_values.items()}
        sharded_key, subkeys = p_split(sharded_key)
        data, aux_data = call_walking_step(subkeys, params, data, aux_data)

        # Logging and saving
        now = time.time()
        if time_start is None:
            time_start = now
        should_log = last_log < now - options.log_interval
        should_save = last_save < now - options.save_interval
        if not should_save and not should_log:
            continue
        all_values_yet = {k: v[: i + 1] for k, v in all_values.items()}

        if options.reweight_ratio > 0.0:
            mean_weights = jnp.mean(all_values_yet.pop("reweighting_weights"))
            all_values_yet = {k: v / mean_weights for k, v in all_values_yet.items()}
        digest = estimator.digest(all_values_yet, state)

        if should_save:
            last_save = now
            checkpoint_mgr.save(i, data, digest, all_values, state, aux_data, metadata)
        # Also log when should save
        last_log = now
        logger.info("Loop %s", i)
        log_digest(i, digest)

    if i == init_step or time_start is None:
        logger.warning("Not enough steps to calculate time per step.")
    else:
        logger.info(
            "Time per step: %.2fs", (time.time() - time_start) / (i - init_step)
        )
    if options.reweight_ratio > 0.0:
        mean_weights = jnp.mean(all_values.pop("reweighting_weights"))
        all_values = {k: v / mean_weights for k, v in all_values.items()}
    digest = estimator.digest(all_values, state)
    log_digest(i, digest)
    checkpoint_mgr.save(i, data, digest, all_values, state, aux_data, metadata)
    return digest, all_values, state


def log_digest(i: int, digest: dict[str, jnp.ndarray]) -> None:
    for k, v in digest.items():
        if v.shape and v.shape[0] == i + 1:
            v, std_v = robust_mean_std(v)
            if v.shape:
                print(k, v, f"std {k}", std_v, sep="\n")
            else:
                print(k + ":", v, "\xb1", std_v)
        else:
            print(k, v, sep="\n")
