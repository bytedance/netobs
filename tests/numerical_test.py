# Copyright 2022-2024 Bytedance Ltd. and/or its affiliates
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import pytest
from jax import numpy as jnp

from netobs.adaptors.simple_hydrogen import SimpleHydrogen
from netobs.evaluate import evaluate_observable
from netobs.helpers.digest import robust_mean
from netobs.observables.density import DensityEstimator
from netobs.observables.energy import EnergyEstimator
from netobs.observables.force import AC, SWCT, Bare
from netobs.observables.wf_change import WFChangeEstimator
from netobs.options import NetObsOptions

pytest.importorskip(
    "jax",
    minversion="0.4.0",
    reason="Different JAX version has different behaviors. Please use JAX 0.4.x",
)


@pytest.fixture
def adaptor():
    return SimpleHydrogen(0.999, [])


@pytest.fixture
def options():
    return NetObsOptions(
        steps=20,
        random_seed=42,
        mcmc_burn_in=100,
        mcmc_steps=10,
        network_restore="2048",
    )


def test_energy(adaptor: SimpleHydrogen, options: NetObsOptions, snapshot: str):
    digest, *_ = evaluate_observable(adaptor, EnergyEstimator, options=options)
    assert digest is not None
    assert "energy" in digest
    energy = robust_mean(digest["energy"])
    assert jnp.allclose(energy, -0.5, rtol=5e-4)
    assert [robust_mean(value) for value in digest.values()] == snapshot


def test_density(adaptor: SimpleHydrogen, options: NetObsOptions, snapshot):
    options.estimator["bins"] = [1, 2, 3]
    options.estimator["range"] = [[-3, 3], [-3, 3], [-3, 3]]
    options.steps = 1
    options.mcmc_burn_in = 0
    *_, state = evaluate_observable(adaptor, DensityEstimator, options=options)
    assert "map" in state
    assert jnp.sum(state["map"]) < 2048
    assert state["map"].shape == (1, 2, 3)
    assert state["map"] == snapshot


def test_bare(adaptor: SimpleHydrogen, options: NetObsOptions, snapshot: str):
    digest, *_ = evaluate_observable(adaptor, Bare, options=options)
    assert digest is not None
    assert "force" in digest
    force = robust_mean(digest["force"])
    assert jnp.allclose(force, 0, atol=0.2)
    assert force == snapshot


def test_antithetic(adaptor: SimpleHydrogen, options: NetObsOptions, snapshot: str):
    options.estimator["r_core"] = 0.5
    digest, *_ = evaluate_observable(adaptor, Bare, options=options)
    assert digest is not None
    assert "force" in digest
    force = robust_mean(digest["force"])
    assert jnp.allclose(force, 0, atol=1e-3)
    assert force == snapshot


def test_antithetic_zb(adaptor: SimpleHydrogen, options: NetObsOptions, snapshot):
    options.estimator["r_core"] = 0.5
    options.estimator["zb"] = True
    digest, *_ = evaluate_observable(adaptor, Bare, options=options)
    assert digest is not None
    assert "force" in digest
    force = robust_mean(digest["force"])
    assert jnp.allclose(force, 0, atol=1e-3)
    assert force == snapshot


def test_zv_noerror(adaptor: SimpleHydrogen, options: NetObsOptions):
    options.mcmc_burn_in = 0
    options.steps = 1
    digest, *_ = evaluate_observable(adaptor, AC, options=options)
    assert digest is not None
    assert "force" in digest


def test_zvzb(adaptor: SimpleHydrogen, options: NetObsOptions, snapshot: str):
    options.estimator["zb"] = True
    digest, *_ = evaluate_observable(adaptor, AC, options=options)
    assert digest is not None
    assert "force" in digest
    assert "force_zv" in digest
    force = robust_mean(digest["force"])
    assert jnp.allclose(force, 0, atol=1e-5)
    force_zv = robust_mean(digest["force_zv"])
    assert jnp.allclose(force_zv, 0, atol=1e-6)
    assert (force, force_zv) == snapshot


def test_swct(adaptor: SimpleHydrogen, options: NetObsOptions, snapshot: str):
    digest, *_ = evaluate_observable(adaptor, SWCT, options=options)
    assert digest is not None
    assert "force" in digest
    assert "force_no_warp" in digest
    force = robust_mean(digest["force"])
    assert jnp.allclose(force, 0, atol=1e-7)
    force_no_warp = robust_mean(digest["force_no_warp"])
    assert jnp.allclose(force_no_warp, 0, atol=1e-3)
    assert (force, force_no_warp) == snapshot


def test_wf_change(adaptor: SimpleHydrogen, options: NetObsOptions, snapshot: str):
    digest, *_ = evaluate_observable(adaptor, WFChangeEstimator, options=options)
    assert digest is not None
    assert "overlap" in digest
    overlap = digest["overlap"]
    assert jnp.allclose(overlap, 1, rtol=1e-4)
    assert overlap == snapshot
