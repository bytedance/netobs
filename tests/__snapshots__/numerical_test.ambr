# serializer version: 1
# name: test_antithetic
  Array([[-0.000805, 0.00113, -0.000496]], dtype=float32)
# ---
# name: test_antithetic_zb
  Array([[-0.000386, -0.001233, -0.001715]], dtype=float32)
# ---
# name: test_bare
  Array([[0.08919, -0.069986, 0.116991]], dtype=float32)
# ---
# name: test_density
  Array([[[75.0, 815.0, 98.0],
          [84.0, 820.0, 98.0]]], dtype=float32)
# ---
# name: test_energy
  list([
    Array(-0.500034, dtype=float32),
    Array(-1.024815, dtype=float32),
    Array(0.524784, dtype=float32),
  ])
# ---
# name: test_swct
  tuple(
    Array([[0.0, 0.0, 0.0]], dtype=float32),
    Array([[9.4e-05, -6.9e-05, 0.000114]], dtype=float32),
  )
# ---
# name: test_wf_change
  Array(0.999996, dtype=float32)
# ---
# name: test_zvzb
  tuple(
    Array([[3e-06, 9e-06, 1e-06]], dtype=float32),
    Array([[0.0, 0.0, 0.0]], dtype=float32),
  )
# ---
